optuna/multi_objective/trial.py

```diff
@@ -0,0 +1,218 @@
+
+CategoricalChoiceType = Union[None, bool, int, float, str]
+
+
+@experimental("1.4.0")
+class MultiObjectiveTrial(object):
+    def __init__(self, trial: Trial):
+        self._trial = trial
+        self._n_objectives = multi_objective.study.MultiObjectiveStudy(trial.study).n_objectives
+
+    def suggest_uniform(self, name: str, low: float, high: float) -> float:
+        return self._trial.suggest_uniform(name, low, high)
+
+    def suggest_loguniform(self, name: str, low: float, high: float) -> float:
+        return self._trial.suggest_loguniform(name, low, high)
+
+    def suggest_discrete_uniform(self, name: str, low: float, high: float, q: float) -> float:
+        return self._trial.suggest_discrete_uniform(name, low, high, q)
+
+    def suggest_int(self, name: str, low: int, high: int) -> int:
+        return self._trial.suggest_int(name, low, high)
+
+    def suggest_categorical(
+        self, name: str, choices: Sequence[CategoricalChoiceType]
+    ) -> CategoricalChoiceType:
+        return self._trial.suggest_categorical(name, choices)
+
+    def report(self, values: List[float], step: int) -> None:
+        if len(values) != self._n_objectives:
+            raise ValueError(
+                "The number of the intermediate values {} at step {} is mismatched with"
+                "the number of the objectives {}.",
+                len(values),
+                step,
+                self._n_objectives,
+            )
+
+        for i, value in enumerate(values):
+            self._trial.report(value, self._n_objectives * (step + 1) + i)
+
+    def _report_complete_values(self, values: List[float]) -> None:
+        if len(values) != self._n_objectives:
+            raise ValueError(
+                "The number of the values {} is mismatched with the number of the objectives {}.",
+                len(values),
+                self._n_objectives,
+            )
+
+        for i, value in enumerate(values):
+            self._trial.report(value, i)
+
+    def set_user_attr(self, key: str, value: Any) -> None:
+        self._trial.set_user_attr(key, value)
+
+    def set_system_attr(self, key: str, value: Any) -> None:
+        self._trial.set_system_attr(key, value)
+
+    @property
+    def params(self) -> Dict[str, Any]:
+        return self._trial.params
+
+    @property
+    def distributions(self) -> Dict[str, BaseDistribution]:
+        return self._trial.distributions
+
+    @property
+    def user_attrs(self) -> Dict[str, Any]:
+        return self._trial.user_attrs
+
+    @property
+    def system_attrs(self) -> Dict[str, Any]:
+        return self._trial.system_attrs
+
+    @property
+    def datetime_start(self) -> Optional[datetime]:
+        return self._trial.datetime_start
+
+    # TODO(ohta): Add `to_single_objective` method.
+    # This method would be helpful to use the existing pruning
+    # integrations for multi-objective optimization.
+
+    @property
+    def _values(self) -> List[Optional[float]]:
```

Comment:
Ditto, how about making this a method as its implementation has non-constant time complexity?

Reply:
Fixed: https://github.com/optuna/optuna/pull/1054/commits/4fadca82957aa7faa2153d048307a34c6c97d66a
Thanks!